Script: Metadata for update context

stu-elastic opened the pull request:
Adds the `metadata()` API call and a Metadata class for the Update context.

There are different metadata available in the update context depending
on whether it is an update or an insert (via upsert).

For update, scripts can read `index`, `id`, `routing`, `version` and `timestamp`.

For insert, scripts can read `index`, `id` and `timestamp`.

Scripts can always read and write the `op` but the available ops are different.

Updates allow 'noop', 'index' and 'delete'.
Inserts allow 'noop' and 'create'.

Refs: #86472

stu-elastic committed:
WIP ingest update metadata

stu-elastic committed:
Create validators

stu-elastic committed:
Merge branch 'master' of github.com:stu-elastic/elasticsearch into ingest_update_meta

stu-elastic committed:
revert "sourceAndMetadata" from "ingestContext"

stu-elastic committed:
revert "getContextMetadata" -> "getMetadata"

stu-elastic committed:
Update javadoc

stu-elastic committed:
rest tests

stu-elastic committed:
Script: Metadata for update context

Adds the metadata() API call and a Metadata class for the Update context

There are different metadata available in the update context depending
on whether it is an update or an insert (via upsert).

For update, scripts can read index, id, routing, version and timestamp.

For insert, scripts can read index, id and timestamp.

Scripts can always read and write the op but the available ops are different.
Updates allow 'noop', 'index' and 'delete'.
Inserts allow 'noop' and 'create'.

Refs: #86472

stu-elastic committed:
add opValidator javadocs

stu-elastic committed:
Add hashcode for WriteableIngestDocumentTests.testFromXContent and leinancy for UpdateRequestTests.testUpdateScript

stu-elastic committed:
create is valid in scripts

stu-elastic committed:
Pass valid ops

stu-elastic committed:
spotless UpdateSourceAndMetadata

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

elasticsearchmachine commented:
Hi @stu-elastic, I've created a changelog YAML for you.

stu-elastic committed:
Update docs/changelog/88333.yaml

elasticmachine commented:
Pinging @elastic/clients-team (Team:Clients)

jdconrad reviewed (commented):
I realized we have talked about this, but after viewing the code I was wondering again what if each context did have its own Metadata and the base class for this took in the Map of validators? The numerous comments in the Metadata class about how something is needed only for one context, but not another seems like it'll be hard to maintain. We also wouldn't have to have methods on the class that don't make sense for a specific context. You still wouldn't have to alias the name because it's part of the allow list specific to each context.

I can see there's a lot of refactoring work as well as you mentioned to me. Since I'm just coming back into this project it's tricky for me to differentiate what's refactoring versus what's added here that's new. Would you be able to split the refactoring into a first PR please?

stu-elastic committed:
WIP

stu-elastic committed:
Split Metadata off off IngestSourceAndMetadata

stu-elastic committed:
Update metadata javadoc

stu-elastic committed:
Add hashcode and equals to make the WritableIngestDocumentTests pass

stu-elastic committed:
Merge branch 'master' of github.com:elastic/elasticsearch into ingest_ctx_map

stu-elastic committed:
keys() -> getKeys(), don't cache entrySet, check metadata in source, return vt validator, doc matcher only checks metadata map

stu-elastic committed:
Return raw map

stu-elastic committed:
Better javadoc for metadata

stu-elastic committed:
Merge branch 'master' of github.com:stu-elastic/elasticsearch into ingest_ctx_map

stu-elastic committed:
Script: Move Map impl from IngestSourceAndMetadata to CtxMap

stu-elastic committed:
Merge branch 'master' of github.com:stu-elastic/elasticsearch into ingest_update_meta

stu-elastic committed:
Merge branch 'ingest_update_meta' of github.com:stu-elastic/elasticsearch into ingest_update_meta

stu-elastic commented:
@jdconrad here's the work to pull Metadata into it's own class, as requested: https://github.com/elastic/elasticsearch/pull/88401

stu-elastic committed:
Script: Configuration driven validation for CtxMap

Adds FieldProperties record that configures how to validate
fields.  Fields have a type, are writeable or read-only, and nullable
or not and may have an additional validation useful for Set/Enum
validation.

Splits IngestMetadata from Metadata in preparation for new Metdata
subclasses.

stu-elastic committed:
Fix hashCode in CtxMap

stu-elastic committed:
Fix hashCode in CtxMap

stu-elastic committed:
Check metadata map in testSimulate

stu-elastic committed:
Merge branch 'ingest_ctx_map' of github.com:stu-elastic/elasticsearch into ingest_ctx_map_field_prop

stu-elastic committed:
Merge branch 'ingest_ctx_map' of github.com:stu-elastic/elasticsearch into ingest_sm_to_ctx_map

stu-elastic committed:
keySet(), Sets.intersection, isAvailable, protected validator & getters

stu-elastic committed:
Merge branch 'ingest_ctx_map' of github.com:stu-elastic/elasticsearch into ingest_sm_to_ctx_map

stu-elastic committed:
Merge branch 'ingest_sm_to_ctx_map' of github.com:stu-elastic/elasticsearch into ingest_ctx_map_field_prop

stu-elastic committed:
Fix unit tests

stu-elastic committed:
FieldProperty javadoc

stu-elastic committed:
Implement getTimestamp() in IngestMetadata

stu-elastic committed:
Use extended validators, add hashcode and equals for Metadata

stu-elastic committed:
_type can be updated

stu-elastic committed:
Fix testExecutePropagateAllMetadataUpdates rather than allow _type to be updateable

stu-elastic committed:
Merge branch 'master' of github.com:stu-elastic/elasticsearch into ingest_ctx_map_field_prop

stu-elastic committed:
Script: Ingest Metadata and CtxMap

stu-elastic committed:
Merge branch 'master' of github.com:stu-elastic/elasticsearch into ingest_ctx_map_field_prop

stu-elastic committed:
Remove getIngestSourceAndMetadata

stu-elastic committed:
Update using CtxMap

stu-elastic committed:
Don't clone properties, add MetadataTests

stu-elastic committed:
Merge branch 'ingest_ctx_map_field_prop' of github.com:stu-elastic/elasticsearch into 220712-update-metadata-0

stu-elastic committed:
_type is nullable

stu-elastic committed:
Merge branch '220712-update-metadata-0' of github.com:stu-elastic/elasticsearch into 220712-update-metadata-0

stu-elastic committed:
Add imports

stu-elastic committed:
Add BulkMetadata for reference

stu-elastic committed:
Merge branch 'master' of github.com:stu-elastic/elasticsearch into 220712-update-metadata-0

stu-elastic committed:
Merge branch '220712-update-metadata-0' of github.com:stu-elastic/elasticsearch into 220712-update-metadata-0

stu-elastic committed:
Remove BulkMetadata

stu-elastic committed:
Move Update and Upsert Metadatas to top level, move op and timestamp to Metadata base class

stu-elastic committed:
Fix timestamp key

stu-elastic committed:
Merge master

stu-elastic committed:
Use CtxMap and Metadata subclasses

stu-elastic committed:
Remove SourceAndMetadataMap implementation

stu-elastic committed:
UpdateScript only needs ctx

stu-elastic committed:
Remove old comment

stu-elastic committed:
Remove <p> in javadoc for UpdateMetadata

stu-elastic commented:
@rjernst and @jdconrad, I've changed this PR to use infrastructure from https://github.com/elastic/elasticsearch/pull/88401 and https://github.com/elastic/elasticsearch/pull/88458.  Can you take another pass?

rjernst commented on modules/lang-painless/src/main/resources/org/elasticsearch/painless/org.elasticsearch.script.update.txt:
```diff
@@ -18,3 +18,17 @@ class org.elasticsearch.painless.api.Json {
+    String getId()
+    String getRouting()
+    long getVersion()
+    String getOp()
```
We discussed having get and set op outside of the metadata on each script type that needs it, as it isn't really metadata about the document. What happened with that?

stu-elastic replied:
If it's on the script, then it conflicts with any variables named `op`, so it would be a backwards compatibility issue.

After thinking about it some more, it seems like a high cost to introduce another concept in addition to `field` and `metadata` for this one-off.  Maybe there's a better way to handle it?

rjernst commented on server/src/main/java/org/elasticsearch/script/UpdateMetadata.java:
```diff
@@ -0,0 +1,113 @@
+        SET_ONCE_LONG
+    );
+
+    protected static BiConsumer<String, String> setValidator(Set<String> valid) {
```
nit: the naming here makes it look like a setter. perhaps `stringSetValidator`?

rjernst commented on server/src/main/java/org/elasticsearch/script/UpdateMetadata.java:
```diff
@@ -0,0 +1,113 @@
+    @Override
+    public String getOp() {
+        String op = super.getOp();
+        if (LEGACY_NOOP_STRING.equals(op) || op == null) {
```
I don't think we can do this translation for the "legacy" value here, it would be breaking. We should look at how/when to deprecate the two noop values outside of this PR.

stu-elastic replied:
I suppose we should also restrict `null` in `setOp` like we're already doing with "none".

stu-elastic replied:
In fact, there shouldn't be validation for `op` at all in the map, as it all gets translated to `UpdateOp.NONE`.

rjernst commented on server/src/main/java/org/elasticsearch/script/UpsertMetadata.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Override
+    public String getRouting() {
+        throw new IllegalArgumentException("routing is unavailable for insert");
```
UnsupportedOperationException would be more typical. We should use the same anywhere else we are failing a get/set method (a user should never see it since we woulnd't have allowed the method on scripting to begin with).

rjernst replied:
> If we split the Update context then a script

I thought we were already planning to do that?

rjernst commented on server/src/main/java/org/elasticsearch/script/UpdateCtxMap.java:
```diff
@@ -0,0 +1,56 @@
+        if (rawSource instanceof Map<?, ?> map) {
+            return (Map<String, Object>) map;
+        }
+        throw new IllegalArgumentException(
```
How could this occur? Since we control the ctx map, wouldn't this be a coding error? Should this be an assert instead?

stu-elastic replied:
An `assert` brings down the node.  Do you think we'd get the same feedback with an Exception without the customer pain? Perhaps `IllegalStateException`.

stu-elastic replied:
Changed to assert

rjernst reviewed (commented):
Thanks for the update Stu! After the other PRs this is much more straightforward and easier to understand. I think there is a backcompat issue to address, but otherwise I just have a few minor comments.

stu-elastic commented on server/src/main/java/org/elasticsearch/script/UpdateMetadata.java:
```diff
@@ -0,0 +1,113 @@
+    @Override
+    public void setOp(String op) {
+        if (LEGACY_NOOP_STRING.equals(op)) {
+            throw new IllegalArgumentException(LEGACY_NOOP_STRING + " is deprecated, use 'noop' instead");
```
It is supported in the map, so there's no bwc issue, what do you think about removing the `" is deprecated"` language, disallow it in set but translate it in `getOp`?

That's asymmetric, but the addition of this class interface seems like the time and place to make that change.

rjernst replied:
Ok, I see now. I think your suggestion is good.

stu-elastic replied:
Ok, keeping it.

stu-elastic committed:
wrap in CtxMap, stringSetValidator, Unsupported instead of IllegalArg in Upsert

stu-elastic committed:
Fix UpsertCtxMapTests expected exceptions

stu-elastic committed:
Assert map in getSource, avoid deprecated language in setOp

jdconrad reviewed (approved):
The new changes look fine to me. Please wait for @rjernst to have another look before committing.

stu-elastic committed:
setOp cannot be null

stu-elastic committed:
op valiation in update is done by the getter/setter

rjernst reviewed (approved):
LGTM.
